### ISSUE TITLE
Map settings: use center+zoom instead of extent

### ISSUE THREAD

**Issue by wonder-sk** [qgis/qgis3.0_api#70]
Currently when setting map view, we use rectangle to set up bounding box of the view. The problem is that the extent still needs to be adjusted to the actual aspect ratio of the view, so we end up with "requested" extent (```extent()```) and "true" visible extent (```visibleExtent()```).

Things get even more confusing when map rotation is considered: in that case the true map view is not aligned with axes and it has to be represented by a quadrilateral (```visiblePolygon()```).

The idea is to change the internal representation of the view from the two extent rectangles (original + adjusted) to center point + zoom (which would be represented as map units per pixel). Extent rectangle or polygon would be just data derived from the above and the "requested" extent could be removed altogether. We would still keep ```setExtent()``` method or similar to compute center+zoom which fits the given area.

This should allow us to simplify API a bit and be more future proof in case we wanted to allow perspective transformations or even true 3D view.

**Comment by NathanW2** [comment]
Will this help to bring back the feature I added to not change the
scale/zoom of the map when opening panels?

**Comment by nyalldawson** [comment]
@NathanW2 I don't think so - my understanding is that the issue with that was that portions of the canvas were hidden. So things like "zoom to selected" would not be able to differentiate the hidden parts of the canvas and would instead result in parts of the selection hidden behind the panels.

**Comment by wonder-sk** [comment]
It should help as map canvas resize handling would not request new extent - the zoom would stay the same and the mechanism could even update the view center point so that the map will not move to the side...

**Comment by strk** [comment]
On Thu, Dec 08, 2016 at 06:34:52PM -0800, Martin Dobias wrote:

> The idea is to change the internal representation of the view from
> the two extent rectangles (original + adjusted) to center point + zoom
> (which would be represented as map units per pixel). Extent rectangle
> or polygon would be just data derived from the above and the "requested"
> extent could be removed altogether. We would still keep ```setExtent()```
> method or similar to compute center+zoom which fits the given area.

Makes a lot of sense to me. Version 3.0 is a good occasion to switch.
